### PR TITLE
HTTP Client.close wrapper for retry and authenticated clients.

### PR DESCRIPTION
**Pull request by isoos**
Handles duplicate exit scope callbacks better (#5644).

### CODE REVIEW

File: app/lib/tool/utils/http.dart [hunk 1]
```diff
@@ -17,14 +17,14 @@ http.Client httpRetryClient({
   http.Client? innerClient,
   int? retries,
 }) {
-  return RetryClient(
```
Comment (jonasfj) [inline app/lib/tool/utils/http.dart:20]: hmm, when do we use this?

We should probably never use `RetryClient` as I don't think it retries if the connection fails when reading the response, I could be wrong...

Anyways, this is orthogonal.

File: app/lib/tool/utils/http.dart [hunk 2]
```diff
@@ -67,3 +67,24 @@ class _AuthenticatedClient extends http.BaseClient {
     super.close();
   }
 }
+
+/// Allows the call of [close] multiple times without throwing an Exception.
+class _MultiCloseClient extends http.BaseClient {
```
Comment (jonasfj) [inline app/lib/tool/utils/http.dart:72]: So the counter argument to doing this is that now you risk one context closing the client used in another context.

When creating an instance of `_AuthenticatedClient` we can specify `_closeInnerClient = false` in the constructor.

File: app/lib/tool/utils/http.dart (hunk 1 above)
Comment (isoos) [reply app/lib/tool/utils/http.dart:20]: We have 8 or 9 references already in pub-dev. While it doesn't retry on certain failures, it was still better than using nothing, so I wouldn't remove it unless we have a better solution.

File: app/lib/tool/utils/http.dart (hunk 2 above)
Comment (isoos) [reply app/lib/tool/utils/http.dart:72]: Hm, I'm not sure if it is the same to not close the inner client. I agree, ideally, this shouldn't be necessary, but this seems to be slightly better than the obscure failure we were getting.

**Comment by isoos** [comment]
Closing as this seems to be not needed.

**Pull request closed without merging**